### PR TITLE
Fix problems in the race condition handling logic in LockJob

### PR DESCRIPTION
**Pull request by johto**
Here's a fix for the problem I talked about in #2.  I also noticed another problem where the code was leaking locks, which I fixed in the same commit, since I didn't consider the intermediate state worth exposing.

For posterity, I wrote [this test code](https://gist.github.com/johto/c247af6e8ddcb7f68701) to test the race condition handling, but I think it's perhaps a bit too crazy to be executed in "go test".  We could includ it as a disabled-by-default test, I guess?

I also didn't do anything to try and prevent the loop from running forever.  I didn't really consider it necessary, but I'm willing to be persuaded the other way.

### CODE REVIEW

File: que.go [hunk 1]
```diff
@@ -216,46 +216,56 @@ func (c *Client) LockJob(queue string) (*Job, error) {
...
+		if err == nil {
+			return &j, nil
+		} else if err == pgx.ErrNoRows {
+			// Encountered job race condition; start over from the beginning.
+			// We're still holding the advisory lock, though, so we need to
+			// release it before resuming.  Otherwise we leak the lock,
+			// eventually causing the server to run out of locks.
+			//
+			// Also swallow the possible error, exactly like in Done.
+			var ok bool
```
Comment (bgentry) [inline que.go]: We probably don't need to declare a new `var ok bool` here, we can just use the one from above.

File: que.go [hunk 2]
```diff
@@ -216,46 +216,56 @@ func (c *Client) LockJob(queue string) (*Job, error) {
...
-		&j.Type,
-		&j.Args,
-		&j.ErrorCount,
-	)
-	if err != nil {
-		c.pool.Release(conn)
-		if err == pgx.ErrNoRows {
-			return nil, nil
+
+	for {
```
Comment (bgentry) [inline que.go]: To eliminate the potential for an infinite loop, can you just change this into a `for i := 0; i < N; i++`, where `N` is some reasonable limit? I don't think it's reasonable for this to retry more than a small handful of times. If that's happening, something is probably wrong.

Then we can just return an exported error of some sort after the `for` loop to cover the repeat race condition edge case.

I understand that just picking a value like 2 or 3 is quite arbitrary, but this is supposed to be a rare edge case so the likelihood of it happening multiple times in a row is quite small. Still, I'd much rather return a rare error than potentially enter into an infinite loop :)

**Comment by bgentry** [comment]
Thanks for this! I agree that your test case is quite a monster :) I haven't taken the time to fully understand the semantics of it, but I'll take your word for it that it's a suitable test for this edge case.

Does it run quickly? I'd hate to have disabled-by-default tests in the suite, as they'd be much more likely to break without notice in the future.

File: que.go (hunk 2 above)
Comment (johto) [inline que.go]: Under high concurrency this race condition isn't *that* rare.  In the tests I ran today, with four workers pulling from a queue, it happened 50,000 times for a queue of 75,000 items.  Admittedly, the workers weren't actually working so they were more likely to hit the race than a normal application probably is.

Perhaps a value of 10 would have workers getting really unlucky under high concurrency.  And like you said, we could export the error, so people who are knowingly in this situation could avoid problems by checking for it.

Comment (bgentry) [inline que.go]: Ok, if it's that common under extremely high concurrency, 10 seems like a perfectly reasonable choice :+1:

**Comment by johto** [comment]
> Does it run quickly? I'd hate to have disabled-by-default tests in the suite, as they'd be much more likely to break without notice in the future.

It adds ~175ms, or 76% into the runtime of "go test".  So it's reasonably fast.

I'm just more worried about how fragile the "lock watcher" goroutine feels.  I wish there was a better way to achieve the same thing, but I can't come up with one.

File: que.go (hunk 1 above)
Comment (johto) [inline que.go]: Good point, will fix.

I'm a bit worried about ignoring the error, though :-(  On one hand it's not clear how to recover from the error, but on the other hand if it happens often enough it could affect unrelated transactions.  It'd have to happen **a lot**, but with the current coding the user wouldn't have any kind of warning beforehand.

I don't have any concrete proposals as to what to do, either, so perhaps things should just be left as-is, at least until someone has a better idea.

Comment (bgentry) [inline que.go]: yeah, I don't have a great idea to improve this either. I'm pretty sure I modeled this after what the ruby version was doing. I mean even if you were to bubble up an error, the user couldn't necessarily do anything to fix it. And it's definitely an internal concern IMO that users shouldn't need to worry about unless it's an irrecoverable condition.

I wonder, if an error is encountered here, what's the likelihood that the conn would be thrown away by the underlying pgx driver?

Comment (johto) [inline que.go]: > yeah, I don't have a great idea to improve this either. I'm pretty sure I modeled this after what the ruby version was doing. I mean even if you were to bubble up an error, the user couldn't necessarily do anything to fix it. And it's definitely an internal concern IMO that users shouldn't need to worry about unless it's an irrecoverable condition.

I agree.

> I wonder, if an error is encountered here, what's the likelihood that the conn would be thrown away by the underlying pgx driver?

I have no idea how any of that works.  I've never used the pgx driver.

**Comment by johto** [comment]
I've pushed a new version addressing the comments here.  I chose not to document ErrAgain with godoc, since maxLockJobAttempts should be large enough that it wouldn't be hit by anyone in practice.

Any thoughts?

**Comment by bgentry** [comment]
I'd prefer to have all exported functions, variables, and constants documented since that's considered a best practice. Otherwise we'll get warnings from `golint` / `go vet`.

**Comment by johto** [comment]
> I'd prefer to have all exported functions, variables, and constants documented since that's considered a best practice.

All right.  I've pushed a new version, where I've fixed this by adding a (somewhat clunky) description for the ErrAgain error.

I also made the test case a lot simpler (and faster!).  I removed two extra lockers, made the synchronization clearer and got rid of the lock observer in favour of using a function to wait until specific backend ids are waiting in the lock queue.

Any thoughts?

**Comment by bgentry** [comment]
Looks great! Thanks! :v:
